File: tools/decode_grid.lua
local base64 = require('src.utils.base64')
local zlib = require('zlib')

local function decodeGrid(tmxPath, layerName)
	local f = io.open(tmxPath, 'r')
	local s = f:read('*a')
	f:close()

	local layerPattern = '<layer[^>]*name="' .. layerName .. '"[^>]*>.-<data%-encoding="base64">%s*([%w%+/=%s]-)%s*</data>'
	local b64 = s:match('name="' .. layerName .. '"[^\n]*--<data encoding="base64">')
	-- crude: find data block after the named layer
	local _, e = s:find('<layer[^>]*name="' .. layerName .. '"')
	local block = s:match('(<data encoding="base64">%s*[%w+/=%s]+%s*</data>)', e or 0)
	if not block then return nil end
	b64 = block:match('<data encoding="base64">%s*([%w+/=%s]+)%s*</data>')
	b64 = b64:gsub('%s+', '')
	local raw = base64.decode(b64)
	local dec = zlib.inflate()(raw)
	return dec
end

local dec = decode('res/map/ll2.tmx', 'ground')
if not dec then print('NO DATA'); return end

local w, h = 36, 23
for y = 0, h - 1 do
	local row = ''
	local yKey = y .. ':'
	for x = 0, w - 1 do
		local idx = y * w * 4 + x * 4 + 1
		local g = string.byte(dec, idx) + string.byte(dec, idx + 1) * 256
		row = row .. (g > 0 and '##' or '..')
	end
	print(string.format('%2d', y) .. ' ' .. row)
end

File: tools/decode_grid.py
import base64 as b64, zlib, re
import sys

def load(path):
    content = open(path).read()
    m = re.search(r'<layer[^>]*name="([^"]+)"[^>]*>(.*?)</layer>', content, re.S)
    name, body = m.group(1), m.group(2)
    dm = re.search(r'<data encoding="base64">\s*([A-Za-z0-9+/=\s]+?)\s*</data>', body, re.S)
    b64str = ''.join(dm.group(1).split())
    raw = b64.b64decode(b64str)
    try:
        dec = zlib.decompress(raw)
    except zlib.error:
        dec = raw
    w, h = 36, 23
    cols = {x: '' for x in range(w)}
    for y in range(h):
        for x in range(w):
            idx = (y * w + x) * 4
            g = int.from_bytes(dec[idx:idx + 4], 'little')
            cols[x] += '#' if ((g & 0x80000000) or g > 0) else '.'
    return cols, w, h

def main():
    cols, w, h = load(sys.argv[1])
    # ladder list: (id, x, y, wpx, hpx) from ll2.tmx ladder layer
    ladders = [(9,608,256,32,96),(10,352,288,32,224),(12,416,96,32,224),
               (13,64,96,32,96),(14,192,96,32,96),(15,192,320,32,128),
               (16,128,192,32,128),(17,0,192,32,128),(18,96,448,32,128),
               (19,224,448,32,136),(20,608,544,32,168),(21,896,544,32,64),
               (22,1024,608,32,96)]
    for lid, x, y, w_, h_ in ladders:
        col = x // 32
        r0, r1 = y // 32, (y + h_) // 32
        print('ladder id=%2d col=%2d rows %2d..%2d (x=%3d y=%3d h=%3d) colMask=%s' %
              (lid, col, r0, r1, x, y, h_, cols[col]))
        solid = [r for r in range(h) if cols[col][r] == '#']
        print('   solid rows at that col:', solid)

main()